[BUG] Panic error when using tm_hcl_expression inside a for loop

**Describe the bug**
I get a panic error when using `tm_hcl_expression` inside a `for` loop. Since it's not possible to use `terramate` variables inside a `for` loop, I have to use `tm_merge` as explained in this issue https://github.com/terramate-io/terramate/issues/1676. 

**To Reproduce**
1. Run command `terramate generate` in the following example (I have made it as simple as possible for the purpose of showing the error):
```
globals {
  role = [
    {
      foo = "bar"
    }
  ]
}


generate_hcl "_main.tf" {

  content {
    main_role = tm_merge({}, {
            for entry in global.role :
                entry.foo => {
                  arn = tm_hcl_expression("data.aws_iam_role.${entry.foo}.arn")
                }
        })
  }
}
```

**Log Output**
```
panic: cannot produce tokens for customdecode.ExpressionVal(&hclsyntax.ScopeTraversalExpr{Traversal:hcl.Traversal{hcl.TraverseRoot{isTraverser:hcl.isTraverser{}, Name:"data", SrcRange:hcl.Range{Filename:"<tm_ternary>", Start:hcl.Pos{Line:1, Column:1, Byte:0}, End:hcl.Pos{Line:1, Column:5, Byte:4}}}, hcl.TraverseAttr{isTraverser:hcl.isTraverser{}, Name:"aws_iam_role", SrcRange:hcl.Range{Filename:"<tm_ternary>", Start:hcl.Pos{Line:1, Column:5, Byte:4}, End:hcl.Pos{Line:1, Column:18, Byte:17}}}, hcl.TraverseAttr{isTraverser:hcl.isTraverser{}, Name:"bar", SrcRange:hcl.Range{Filename:"<tm_ternary>", Start:hcl.Pos{Line:1, Column:18, Byte:17}, End:hcl.Pos{Line:1, Column:22, Byte:21}}}, hcl.TraverseAttr{isTraverser:hcl.isTraverser{}, Name:"arn", SrcRange:hcl.Range{Filename:"<tm_ternary>", Start:hcl.Pos{Line:1, Column:22, Byte:21}, End:hcl.Pos{Line:1, Column:26, Byte:25}}}}, SrcRange:hcl.Range{Filename:"<tm_ternary>", Start:hcl.Pos{Line:1, Column:1, Byte:0}, End:hcl.Pos{Line:1, Column:26, Byte:25}}})

goroutine 1 [running]:
github.com/hashicorp/hcl/v2/hclwrite.appendTokensForValue({{{0x101c1b640, 0x140004e59b0}}, {0x1018e0160, 0x1400080f250}}, {0x1400088db00, 0x8, 0x8})
        github.com/hashicorp/hcl/v2@v2.20.1/hclwrite/generate.go:301 +0xde8
github.com/hashicorp/hcl/v2/hclwrite.appendTokensForValue({{{0x101c1b100, 0x1400080f260}}, {0x1019bc9c0, 0x14000b97c50}}, {0x140008b6840, 0x4, 0x4})
        github.com/hashicorp/hcl/v2@v2.20.1/hclwrite/generate.go:287 +0xc14
github.com/hashicorp/hcl/v2/hclwrite.appendTokensForValue({{{0x101c1b100, 0x1400080f340}}, {0x1019bc9c0, 0x1400067b020}}, {0x0, 0x0, 0x0})
        github.com/hashicorp/hcl/v2@v2.20.1/hclwrite/generate.go:287 +0xc14
github.com/hashicorp/hcl/v2/hclwrite.TokensForValue({{{0x101c1b100?, 0x1400080f340?}}, {0x1019bc9c0?, 0x1400067b020?}})
        github.com/hashicorp/hcl/v2@v2.20.1/hclwrite/generate.go:27 +0x38
github.com/terramate-io/terramate/hcl/ast.TokensForValue({{{0x101c1b100?, 0x1400080f340?}}, {0x1019bc9c0?, 0x1400067b020?}})
        github.com/terramate-io/terramate/hcl/ast/expr.go:42 +0x110
github.com/terramate-io/terramate/hcl/ast.(*tokenBuilder).literalTokens(...)
        github.com/terramate-io/terramate/hcl/ast/expr.go:101
github.com/terramate-io/terramate/hcl/ast.(*tokenBuilder).build(0x140007c8e30, {0x101c1b2f8?, 0x140001b2c60?})
        github.com/terramate-io/terramate/hcl/ast/expr.go:62 +0xdc
github.com/terramate-io/terramate/hcl/ast.tokensForExpression(...)
        github.com/terramate-io/terramate/hcl/ast/expr.go:47
github.com/terramate-io/terramate/hcl/ast.TokensForExpression({0x101c1b2f8?, 0x140001b2c60?})
        github.com/terramate-io/terramate/hcl/ast/expr.go:28 +0x40
github.com/terramate-io/terramate/generate/genhcl.copyBody(0x140008b53e0, 0x14000b94d10, {0x101c1b8c0, 0x1400080ef60})
        github.com/terramate-io/terramate/generate/genhcl/genhcl.go:429 +0x240
github.com/terramate-io/terramate/generate/genhcl.Load(0x140000d24c8, 0x140008ae1c0, 0x14000ac7e90, {{0x101206805?, 0x10120320f?}}, 0x14000112de0)
        github.com/terramate-io/terramate/generate/genhcl/genhcl.go:340 +0xe40
github.com/terramate-io/terramate/generate.loadStackCodeCfgs(0x140000d24c8, 0x140008ae1c0, 0x14000ac7e90, {{0x101206805?, 0x0?}}, 0x14000112de0)
        github.com/terramate-io/terramate/generate/generate.go:1198 +0x94
github.com/terramate-io/terramate/generate.doStackGeneration(0x140000d24c8, {{0x101206805?, 0x0?}}, 0x14000112de0)
        github.com/terramate-io/terramate/generate/generate.go:216 +0x3e0
github.com/terramate-io/terramate/generate.Do(0x140000d24c8, {{0x101206805?, 0x0?}}, 0x1400055e660?)
        github.com/terramate-io/terramate/generate/generate.go:184 +0x44
github.com/terramate-io/terramate/cmd/terramate/cli.(*cli).gencodeWithVendor(0x140000d2408)
        github.com/terramate-io/terramate/cmd/terramate/cli/cli.go:1059 +0x114
github.com/terramate-io/terramate/cmd/terramate/cli.(*cli).generate(0x140000d2408)
        github.com/terramate-io/terramate/cmd/terramate/cli/cli.go:1016 +0x60
github.com/terramate-io/terramate/cmd/terramate/cli.(*cli).run(0x140000d2408)
        github.com/terramate-io/terramate/cmd/terramate/cli/cli.go:654 +0x4c0
github.com/terramate-io/terramate/cmd/terramate/cli.Exec({0x101202fb7, 0x5}, {0x14000148010, 0x1, 0x1}, {0x101c091a8, 0x14000128040}, {0x101c091c8, 0x14000128048}, {0x101c091c8, ...})
        github.com/terramate-io/terramate/cmd/terramate/cli/cli.go:340 +0xb0
main.main()
        github.com/terramate-io/terramate/cmd/terramate/main.go:20 +0x94
```
**Environment (please complete the following information):**
 - OS:  Mac 14.4.1
 - Git version: 2.39.3 (Apple Git-146)
 - Terramate Version: 0.5.3 <-> 0.8.1

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,6 +22,14 @@ Given a version number `MAJOR.MINOR.PATCH`, we increment the:
 
 ## Unreleased
 
+### Added
+
+- Add support for `[for ...]` and `{for ...}` expressions containing Terramate variables and functions inside the `generate_hcl.content` block.
+
+### Changed
+
+- (**BREAKING CHANGE**) The format of the generated code may change while being still semantically the same as before. This change is marked as "breaking", because this may trigger change detection on files where the formatting changes.
+
 ## v0.8.4
 
 ### Fixed

--- a/cmd/terramate/cli/cli.go
+++ b/cmd/terramate/cli/cli.go
@@ -2048,7 +2048,7 @@ func (c *cli) partialEval() {
 		if err != nil {
 			fatalWithDetails(err, "unable to parse expression")
 		}
-		newexpr, err := ctx.PartialEval(expr)
+		newexpr, _, err := ctx.PartialEval(expr)
 		if err != nil {
 			fatalWithDetails(err, "partial eval %q", exprStr)
 		}

--- a/cmd/terramate/e2etests/core/exp_eval_test.go
+++ b/cmd/terramate/e2etests/core/exp_eval_test.go
@@ -44,7 +44,7 @@ func TestExpEval(t *testing.T) {
 				Stdout: addnl("false"),
 			},
 			wantPartial: RunExpected{
-				Stdout: addnl("true && false"),
+				Stdout: addnl("false"),
 			},
 		},
 		{
@@ -54,7 +54,7 @@ func TestExpEval(t *testing.T) {
 				Stdout: addnl("[1, 1, 2, 3, 5, 8]"),
 			},
 			wantPartial: RunExpected{
-				Stdout: addnl("[1, 1 + 0, 1 + 1, 1 + 2, 3 + 2, 5 + 3]"),
+				Stdout: addnl("[1, 1, 2, 3, 5, 8]"),
 			},
 		},
 		{
@@ -102,7 +102,7 @@ func TestExpEval(t *testing.T) {
 				Stdout: addnl(`50`),
 			},
 			wantPartial: RunExpected{
-				Stdout: addnl(`49 + 1`),
+				Stdout: addnl(`50`),
 			},
 		},
 		{
@@ -121,7 +121,7 @@ func TestExpEval(t *testing.T) {
 				Stdout: addnl(`20`),
 			},
 			wantPartial: RunExpected{
-				Stdout: addnl(`10 + 10`),
+				Stdout: addnl(`20`),
 			},
 		},
 		{

--- a/cmd/terramate/e2etests/core/generate_test.go
+++ b/cmd/terramate/e2etests/core/generate_test.go
@@ -330,6 +330,76 @@ Hint: '+', '~' and '-' mean the file was created, changed and deleted, respectiv
 				},
 			},
 		},
+		{
+			name: "generate_hcl with for-loops and tm_hcl_expression",
+			layout: []string{
+				"s:stack1",
+				"s:stack2",
+			},
+			files: []file{
+				{
+					path: p("/config.tm"),
+					body: Doc(
+						Globals(
+							Expr("numbers", `[1, 2, 3]`),
+							Expr("strings", `["terramate", "is", "fun"]`),
+							Expr("partials_config", `{
+								aws   = "aws.vpc.id"
+								azure = "azure.thing"
+							}`),
+						),
+						GenerateHCL(
+							Labels("example.hcl"),
+							Content(
+								Expr("square_numbers", `[for n in global.numbers : n*n]`),
+								Expr("upper_strings", `[for s in global.strings : tm_upper(s)]`),
+								Expr("config", `{for k, v in global.partials_config : k => tm_hcl_expression("${terramate.stack.name}.${v}")}`),
+							),
+						),
+					),
+				},
+			},
+			want: want{
+				run: RunExpected{
+					Stdout: `Code generation report
+
+Successes:
+
+- /stack1
+	[+] example.hcl
+
+- /stack2
+	[+] example.hcl
+
+Hint: '+', '~' and '-' mean the file was created, changed and deleted, respectively.
+`,
+				},
+				files: []file{
+					{
+						path: p("/stack1/example.hcl"),
+						body: Doc(
+							Expr("config", `{
+								aws   = stack1.aws.vpc.id
+								azure = stack1.azure.thing
+							}`),
+							Expr("square_numbers", `[1, 4, 9]`),
+							Expr("upper_strings", `["TERRAMATE", "IS", "FUN"]`),
+						),
+					},
+					{
+						path: p("/stack2/example.hcl"),
+						body: Doc(
+							Expr("config", `{
+								aws   = stack2.aws.vpc.id
+								azure = stack2.azure.thing
+							}`),
+							Expr("square_numbers", `[1, 4, 9]`),
+							Expr("upper_strings", `["TERRAMATE", "IS", "FUN"]`),
+						),
+					},
+				},
+			},
+		},
 	}
 
 	for _, tc := range tcases {

--- a/generate/genhcl/dynamic_test.go
+++ b/generate/genhcl/dynamic_test.go
@@ -1037,9 +1037,9 @@ func TestGenerateHCLDynamic(t *testing.T) {
 						condition: true,
 						body: Doc(
 							Block("references",
+								Bool("GLOBALKEY", true),
 								Bool("globalkey", true),
 								Bool("stack", true),
-								Bool("GLOBALKEY", true),
 								Bool("test", true),
 							),
 						),

--- a/generate/genhcl/genhcl.go
+++ b/generate/genhcl/genhcl.go
@@ -421,7 +421,7 @@ func copyBody(dest *hclwrite.Body, src *hclsyntax.Body, eval hcl.Evaluator) erro
 			Expression: attr.Expr.(hclsyntax.Expression),
 		}
 
-		newexpr, err := eval.PartialEval(expr)
+		newexpr, _, err := eval.PartialEval(expr)
 		if err != nil {
 			return errors.E(err, attr.Expr.Range())
 		}
@@ -482,7 +482,7 @@ func appendDynamicBlock(
 
 	attributeNames := map[string]struct{}{}
 	if attrs.attributes != nil {
-		attrsExpr, err := evaluator.PartialEval(attrs.attributes.Expr)
+		attrsExpr, _, err := evaluator.PartialEval(attrs.attributes.Expr)
 		if err != nil {
 			return errors.E(ErrDynamicAttrsEval, err, attrs.attributes.Range())
 		}
@@ -526,7 +526,7 @@ func appendDynamicBlock(
 						keyVal.Type().FriendlyName())
 				}
 
-				valExpr, err := evaluator.PartialEval(item.ValueExpr)
+				valExpr, _, err := evaluator.PartialEval(item.ValueExpr)
 				if err != nil {
 					return errors.E(
 						ErrDynamicAttrsEval,

--- a/generate/genhcl/genhcl_test.go
+++ b/generate/genhcl/genhcl_test.go
@@ -416,9 +416,9 @@ func TestGenerateHCL(t *testing.T) {
 							Bool("bool", true),
 							Number("number", 777),
 							EvalExpr(t, "obj", `{
-								string = "string"
-								number = 777
 								bool   = true
+								number = 777
+								string = "string"
 							}`),
 							Str("string", "string"),
 						),
@@ -487,9 +487,9 @@ func TestGenerateHCL(t *testing.T) {
 						condition: true,
 						body: Block("testblock2",
 							EvalExpr(t, "obj", `{
-								string = "string"
-								number = 777
 								bool   = true
+								number = 777
+								string = "string"
 							}`),
 						),
 					},
@@ -557,9 +557,9 @@ func TestGenerateHCL(t *testing.T) {
 						condition: true,
 						body: Block("testblock2",
 							EvalExpr(t, "obj", `{
-								string = "string"
-								number = 777
 								bool   = true
+								number = 777
+								string = "string"
 							}`),
 						),
 					},
@@ -659,9 +659,9 @@ func TestGenerateHCL(t *testing.T) {
 								Number("number", 777),
 								Block("block3",
 									EvalExpr(t, "obj", `{
-										string = "string"
-										number = 777
 										bool   = true
+										number = 777
+										string = "string"
 									}`),
 									Str("string", "string"),
 								),
@@ -781,9 +781,9 @@ func TestGenerateHCL(t *testing.T) {
 						condition: true,
 						body: Block("on_parent_block",
 							EvalExpr(t, "obj", `{
-								string = "string"
-								number = 777
 								bool   = true
+								number = 777
+								string = "string"
 							}`),
 						),
 					},
@@ -1449,9 +1449,9 @@ func TestGenerateHCL(t *testing.T) {
 							Bool("bool", true),
 							Number("number", 777),
 							EvalExpr(t, "obj", `{
-								string = "string"
-								number = 777
 								bool   = true
+								number = 777
+								string = "string"
 							}`),
 							Str("string", "string"),
 						),
@@ -1504,9 +1504,9 @@ func TestGenerateHCL(t *testing.T) {
 							Bool("bool", true),
 							Number("number", 777),
 							EvalExpr(t, "obj", `{
-								string = "string"
-								number = 777
 								bool   = true
+								number = 777
+								string = "string"
 							}`),
 							Str("string", "string"),
 						),
@@ -1555,9 +1555,9 @@ func TestGenerateHCL(t *testing.T) {
 							Bool("bool", true),
 							Number("number", 777),
 							EvalExpr(t, "obj", `{
-								string = "string"
-								number = 777
 								bool   = true
+								number = 777
+								string = "string"
 							}`),
 							Str("string", "string"),
 						),
@@ -1827,9 +1827,9 @@ func TestGenHCLTmGen(t *testing.T) {
 							Bool("bool", true),
 							Number("number", 777),
 							EvalExpr(t, "obj", `{
-								string = "string"
-								number = 777
 								bool   = true
+								number = 777
+								string = "string"
 							}`),
 							Str("stack_name", "stack"),
 							Str("string", "string"),

--- a/generate/genhcl/partial_eval_test.go
+++ b/generate/genhcl/partial_eval_test.go
@@ -120,7 +120,7 @@ func TestPartialEval(t *testing.T) {
 			),
 			want: Doc(
 				Expr("obj", `{
-					"key" = "value"	
+					key = "value"	
 				}`),
 			),
 		},
@@ -139,7 +139,7 @@ func TestPartialEval(t *testing.T) {
 			),
 			want: Doc(
 				Expr("obj", `{
-					"KEY" = "value"	
+					KEY = "value"	
 				}`),
 			),
 		},
@@ -616,7 +616,7 @@ func TestPartialEval(t *testing.T) {
 				Expr("var", `1 + 1`),
 			),
 			want: Doc(
-				Expr("var", `1 + 1`),
+				Expr("var", `2`),
 			),
 		},
 		{
@@ -629,7 +629,7 @@ func TestPartialEval(t *testing.T) {
 			),
 		},
 		{
-			name: "plus expression evaluated",
+			name: "plus expression with strings evaluated -- fails",
 			globals: Doc(
 				Globals(
 					Str("a", "hello"),
@@ -639,32 +639,45 @@ func TestPartialEval(t *testing.T) {
 			config: Doc(
 				Expr("var", `tm_upper(global.a) + tm_upper(global.b)`),
 			),
+			wantErr: errors.E(eval.ErrPartial),
+		},
+		{
+			name: "plus expression with numbers evaluated",
+			globals: Doc(
+				Globals(
+					Number("a", 2),
+					Number("b", 2),
+				),
+			),
+			config: Doc(
+				Expr("var", `global.a*global.a + global.b*global.b`),
+			),
 			want: Doc(
-				Expr("var", `"HELLO" + "WORLD"`),
+				Expr("var", `8`),
 			),
 		},
 		{
 			name: "plus expression evaluated advanced",
 			globals: Doc(
 				Globals(
-					Str("a", "hello"),
-					Str("b", "world"),
+					Number("a", 100),
+					Number("b", 200),
 				),
 			),
 			config: Doc(
-				Expr("var", `tm_lower(tm_upper(global.a)) + tm_lower(tm_upper(global.b))`),
+				Expr("var", `tm_max(global.a, global.b) + tm_min(global.b, global.a)`),
 			),
 			want: Doc(
-				Expr("var", `"hello" + "world"`),
+				Number("var", 300),
 			),
 		},
 		{
 			name: "basic minus expression",
 			config: Doc(
-				Expr("var", `1 + 1`),
+				Expr("var", `1 - 1`),
 			),
 			want: Doc(
-				Expr("var", `1 + 1`),
+				Number("var", 0),
 			),
 		},
 		{
@@ -673,7 +686,7 @@ func TestPartialEval(t *testing.T) {
 				Expr("var", `1 == 1 ? 0 : 1`),
 			),
 			want: Doc(
-				Expr("var", `1 == 1 ? 0 : 1`),
+				Number("var", 0),
 			),
 		},
 		{
@@ -1034,24 +1047,28 @@ func TestPartialEval(t *testing.T) {
 			),
 		},
 		{
-			name: "list for loop with global reference fails",
+			name: "list for loop with global reference",
 			globals: Globals(
 				Expr("list", `["a", "b", "c"]`),
 			),
 			config: Doc(
 				Expr("list", `[for k in global.list : k]`),
 			),
-			wantErr: errors.E(eval.ErrForExprDisallowEval),
+			want: Doc(
+				Expr("list", `["a", "b", "c"]`),
+			),
 		},
 		{
-			name: "obj for loop with global reference fails",
+			name: "obj for loop with global reference",
 			globals: Globals(
 				Expr("obj", `{ a = 1}`),
 			),
 			config: Doc(
 				Expr("obj", `[for k in global.obj : k]`),
 			),
-			wantErr: errors.E(eval.ErrForExprDisallowEval),
+			want: Doc(
+				Expr("obj", `[1]`),
+			),
 		},
 		{
 			name: "[for in from [for list with global references",
@@ -1061,7 +1078,9 @@ func TestPartialEval(t *testing.T) {
 			config: Doc(
 				Expr("obj", `[for s in [for s in global.list : s] : s]`),
 			),
-			wantErr: errors.E(eval.ErrForExprDisallowEval),
+			want: Doc(
+				Expr("obj", `["a", "b", "c"]`),
+			),
 		},
 		{
 			name: "mixing {for and [for",
@@ -1075,10 +1094,10 @@ func TestPartialEval(t *testing.T) {
 		{
 			name: "unary operation !",
 			config: Doc(
-				Expr("num", "!0"),
+				Expr("b", "!false"),
 			),
 			want: Doc(
-				Expr("num", "!0"),
+				Expr("b", "true"),
 			),
 		},
 		{
@@ -1127,14 +1146,14 @@ func TestPartialEval(t *testing.T) {
 		{
 			name: "conditional globals evaluation",
 			globals: Globals(
-				Str("domain", "mineiros.io"),
+				Str("domain", "terramate.io"),
 				Bool("exists", true),
 			),
 			config: Doc(
 				Expr("a", `global.exists ? global.domain : "example.com"`),
 			),
 			want: Doc(
-				Expr("a", `true ? "mineiros.io" : "example.com"`),
+				Expr("a", `"terramate.io"`),
 			),
 		},
 		{
@@ -1483,6 +1502,140 @@ EOT
 				Expr("a", `{
 					a = b
 				}`),
+			),
+		},
+		{
+			name: "tm_hcl_expression with for-object loops",
+			globals: Globals(
+				Expr("role", `[
+                   {
+					  foo = "bar"
+				   },
+				   {
+					  foo = "baz"
+				   }
+				]`),
+			),
+			config: Doc(
+				Expr("main_role", `{
+					for k, v in global.role : v.foo => tm_hcl_expression("data.aws_iam_roles[${k}].${v.foo}.arn")
+				}`),
+			),
+			want: Doc(
+				Expr("main_role", `{
+					bar = data.aws_iam_roles[0].bar.arn
+					baz = data.aws_iam_roles[1].baz.arn
+				}`),
+			),
+		},
+		{
+			// https://github.com/terramate-io/terramate/issues/1689
+			name: "regression test: tm_hcl_expression with for-object loops + tm_merge",
+			globals: Globals(
+				Expr("role", `[
+                   {
+					  foo = "bar"
+				   },
+				   {
+					  foo = "baz"
+				   }
+				]`),
+			),
+			config: Doc(
+				Expr("main_role", `tm_merge({}, {
+					for k, v in global.role : v.foo => tm_hcl_expression("data.aws_iam_roles[${k}].${v.foo}.arn")
+				})`),
+			),
+			want: Doc(
+				Expr("main_role", `{
+					bar = data.aws_iam_roles[0].bar.arn
+					baz = data.aws_iam_roles[1].baz.arn
+				}`),
+			),
+		},
+		{
+			// https://github.com/terramate-io/terramate/issues/1676
+			name: "regression test: tm_hcl_expression with tm_dynamic",
+			globals: Globals(
+				Labels("envs"),
+				Expr("sandbox", `{
+					required = true
+					reviewers = [
+						"a-reviewer"
+					]
+				}`),
+				Expr("staging", `{
+					required = false
+					waiter = 10
+					reviewers = [
+						"another-reviewer",
+						"a-reviewer"
+					]
+				}`),
+				Expr("production", `{
+					waiter = 100
+					reviewers = [
+						"another-reviewer"
+					]
+				}`),
+			),
+			config: Doc(
+				TmDynamic(
+					Labels("resource"),
+					Expr("for_each", `global.envs`),
+					Expr("iterator", "env"),
+					Expr("labels", `["github_environment", env.key]`),
+					Expr("attributes", `{ for k, v in env.value : k => v if !tm_can(tm_keys(v)) }`),
+				),
+			),
+			want: Doc(
+				Block("resource",
+					Labels("github_environment", "production"),
+					Expr("reviewers", `[
+						"another-reviewer",
+					  ]`),
+					Number("waiter", 100),
+				),
+				Block("resource",
+					Labels("github_environment", "sandbox"),
+					Bool("required", true),
+					Expr("reviewers", `[
+						"a-reviewer",
+					  ]`),
+				),
+				Block("resource",
+					Labels("github_environment", "staging"),
+					Bool("required", false),
+					Expr("reviewers", `[
+						"another-reviewer",
+						"a-reviewer",
+					  ]`),
+					Number("waiter", 10),
+				),
+			),
+		},
+		{
+			name: "tm_hcl_expression with for-list loops",
+			globals: Globals(
+				Expr("role", `[
+                   {
+					  foo = "bar"
+				   },
+				   {
+					  foo = "baz"
+				   }
+				]`),
+			),
+			config: Doc(
+				Expr("main_role", `[
+					for k, v in global.role : tm_hcl_expression("data.whatever[${k}].other.${v.foo}")
+				]`),
+			),
+			want: Doc(
+				Expr("main_role", `[
+					data.whatever[0].other.bar,
+					data.whatever[1].other.baz,
+				]`),
 			),
 		},
 		{

--- a/hcl/ast/expr.go
+++ b/hcl/ast/expr.go
@@ -5,13 +5,12 @@ package ast
 
 import (
 	"fmt"
+	"math/big"
 
 	"github.com/hashicorp/hcl/v2"
-	"github.com/hashicorp/hcl/v2/ext/customdecode"
 	"github.com/hashicorp/hcl/v2/hclsyntax"
 	"github.com/hashicorp/hcl/v2/hclwrite"
 	"github.com/terramate-io/terramate/errors"
-	"github.com/zclconf/go-cty/cty"
 )
 
 // ParseExpression parses the expression str.
@@ -31,20 +30,9 @@ func TokensForExpression(expr hcl.Expression) hclwrite.Tokens {
 	return tokens
 }
 
-// TokensForValue returns the tokens for the provided value.
-func TokensForValue(value cty.Value) hclwrite.Tokens {
-	if value.Type() == customdecode.ExpressionClosureType {
-		closureExpr := value.EncapsulatedValue().(*customdecode.ExpressionClosure)
-		return TokensForExpression(closureExpr.Expression)
-	} else if value.Type() == customdecode.ExpressionType {
-		return TokensForExpression(customdecode.ExpressionFromVal(value))
-	}
-	return hclwrite.TokensForValue(value)
-}
-
 func tokensForExpression(expr hcl.Expression) hclwrite.Tokens {
 	builder := tokenBuilder{}
-	builder.build(expr)
+	builder.fromExpr(expr)
 	return builder.tokens
 }
 
@@ -56,7 +44,7 @@ func (builder *tokenBuilder) add(tokens ...*hclwrite.Token) {
 	builder.tokens = append(builder.tokens, tokens...)
 }
 
-func (builder *tokenBuilder) build(expr hcl.Expression) {
+func (builder *tokenBuilder) fromExpr(expr hcl.Expression) {
 	switch e := expr.(type) {
 	case *hclsyntax.LiteralValueExpr:
 		builder.literalTokens(e)
@@ -197,12 +185,12 @@ func (builder *tokenBuilder) templateTokens(tmpl *hclsyntax.TemplateExpr) {
 
 func (builder *tokenBuilder) templateWrapTokens(tmpl *hclsyntax.TemplateWrapExpr) {
 	builder.add(oquote(), interpBegin())
-	builder.build(tmpl.Wrapped)
+	builder.fromExpr(tmpl.Wrapped)
 	builder.add(interpEnd(), cquote())
 }
 
 func (builder *tokenBuilder) binOpTokens(binop *hclsyntax.BinaryOpExpr) {
-	builder.build(binop.LHS)
+	builder.fromExpr(binop.LHS)
 	var op *hclwrite.Token
 	switch binop.Op {
 	case hclsyntax.OpAdd:
@@ -237,7 +225,7 @@ func (builder *tokenBuilder) binOpTokens(binop *hclsyntax.BinaryOpExpr) {
 	op.SpacesBefore = 1
 	builder.add(op)
 	nexttok := len(builder.tokens)
-	builder.build(binop.RHS)
+	builder.fromExpr(binop.RHS)
 	builder.tokens[nexttok].SpacesBefore = 1
 }
 
@@ -250,19 +238,19 @@ func (builder *tokenBuilder) unaryOpTokens(unary *hclsyntax.UnaryOpExpr) {
 	default:
 		panic(fmt.Sprintf("value %+v is unexpected\n", unary.Op))
 	}
-	builder.build(unary.Val)
+	builder.fromExpr(unary.Val)
 }
 
 func (builder *tokenBuilder) parenExprTokens(parenExpr *hclsyntax.ParenthesesExpr) {
 	builder.add(oparen())
-	builder.build(parenExpr.Expression)
+	builder.fromExpr(parenExpr.Expression)
 	builder.add(cparen())
 }
 
 func (builder *tokenBuilder) tupleTokens(tuple *hclsyntax.TupleConsExpr) {
 	builder.add(obrack())
 	for i, expr := range tuple.Exprs {
-		builder.build(expr)
+		builder.fromExpr(expr)
 		if i+1 != len(tuple.Exprs) {
 			builder.add(comma())
 		}
@@ -276,10 +264,10 @@ func (builder *tokenBuilder) objectTokens(obj *hclsyntax.ObjectConsExpr) {
 		builder.add(nl())
 	}
 	for _, item := range obj.Items {
-		builder.build(item.KeyExpr)
+		builder.fromExpr(item.KeyExpr)
 		builder.add(assign(1))
 		nexttok := len(builder.tokens)
-		builder.build(item.ValueExpr)
+		builder.fromExpr(item.ValueExpr)
 		builder.tokens[nexttok].SpacesBefore = 1
 		builder.add(nl())
 	}
@@ -288,13 +276,13 @@ func (builder *tokenBuilder) objectTokens(obj *hclsyntax.ObjectConsExpr) {
 
 func (builder *tokenBuilder) objectKeyTokens(key *hclsyntax.ObjectConsKeyExpr) {
 	// TODO(i4k): review the case for key.ForceNonLiteral = true|false
-	builder.build(key.Wrapped)
+	builder.fromExpr(key.Wrapped)
 }
 
 func (builder *tokenBuilder) funcallTokens(fn *hclsyntax.FunctionCallExpr) {
 	builder.add(ident(fn.Name, 0), oparen())
 	for i, expr := range fn.Args {
-		builder.build(expr)
+		builder.fromExpr(expr)
 		if i+1 != len(fn.Args) {
 			builder.add(comma())
 		}
@@ -306,11 +294,11 @@ func (builder *tokenBuilder) funcallTokens(fn *hclsyntax.FunctionCallExpr) {
 }
 
 func (builder *tokenBuilder) conditionalTokens(cond *hclsyntax.ConditionalExpr) {
-	builder.build(cond.Condition)
+	builder.fromExpr(cond.Condition)
 	builder.add(question())
-	builder.build(cond.TrueResult)
+	builder.fromExpr(cond.TrueResult)
 	builder.add(colon())
-	builder.build(cond.FalseResult)
+	builder.fromExpr(cond.FalseResult)
 }
 
 func (builder *tokenBuilder) forExprTokens(forExpr *hclsyntax.ForExpr) {
@@ -339,11 +327,11 @@ func (builder *tokenBuilder) forExprTokens(forExpr *hclsyntax.ForExpr) {
 	builder.add(in...)
 	builder.add(colon())
 	if forExpr.KeyExpr != nil {
-		builder.build(forExpr.KeyExpr)
+		builder.fromExpr(forExpr.KeyExpr)
 		builder.add(arrow())
-		builder.build(forExpr.ValExpr)
+		builder.fromExpr(forExpr.ValExpr)
 	} else {
-		builder.build(forExpr.ValExpr)
+		builder.fromExpr(forExpr.ValExpr)
 	}
 	if forExpr.Group {
 		builder.add(ellipsis())
@@ -351,25 +339,25 @@ func (builder *tokenBuilder) forExprTokens(forExpr *hclsyntax.ForExpr) {
 	if forExpr.CondExpr != nil {
 		builder.add(ident("if", 1))
 		nexttok := len(builder.tokens)
-		builder.build(forExpr.CondExpr)
+		builder.fromExpr(forExpr.CondExpr)
 		builder.tokens[nexttok].SpacesBefore = 1
 	}
 	builder.add(end)
 }
 
 func (builder *tokenBuilder) indexTokens(index *hclsyntax.IndexExpr) {
-	builder.build(index.Collection)
+	builder.fromExpr(index.Collection)
 	builder.add(obrack())
-	builder.build(index.Key)
+	builder.fromExpr(index.Key)
 	builder.add(cbrack())
 }
 
 func (builder *tokenBuilder) splatTokens(splat *hclsyntax.SplatExpr) {
-	builder.build(splat.Source)
+	builder.fromExpr(splat.Source)
 	builder.add(obrack())
 	builder.add(star())
 	builder.add(cbrack())
-	builder.build(splat.Each)
+	builder.fromExpr(splat.Each)
 }
 
 func (builder *tokenBuilder) scopeTraversalTokens(scope *hclsyntax.ScopeTraversalExpr) {
@@ -397,7 +385,7 @@ func (builder *tokenBuilder) traversalTokens(traversals hcl.Traversal) {
 }
 
 func (builder *tokenBuilder) relTraversalTokens(traversal *hclsyntax.RelativeTraversalExpr) {
-	builder.build(traversal.Source)
+	builder.fromExpr(traversal.Source)
 	builder.traversalTokens(traversal.Traversal)
 }
 
@@ -484,6 +472,14 @@ func renderString(bytes []byte) []byte {
 		}
 	}
 	return out
+}
+
+func num(n *big.Float) *hclwrite.Token {
+	srcStr := n.Text('f', -1)
+	return &hclwrite.Token{
+		Type:  hclsyntax.TokenNumberLit,
+		Bytes: []byte(srcStr),
+	}
 }
 
 func obrace() *hclwrite.Token {

--- a/hcl/ast/value.go
+++ b/hcl/ast/value.go
@@ -1,0 +1,134 @@
+// Copyright 2024 Terramate GmbH
+// SPDX-License-Identifier: MPL-2.0
+
+package ast
+
+import (
+	"fmt"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/hashicorp/hcl/v2/ext/customdecode"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/terramate-io/terramate/errors"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// TokensForValue returns the tokens for the provided value.
+func TokensForValue(value cty.Value) hclwrite.Tokens {
+	builder := tokenBuilder{}
+	builder.fromValue(value)
+	return builder.tokens
+}
+
+func (builder *tokenBuilder) fromValue(val cty.Value) {
+	switch typ := val.Type(); {
+	case typ == customdecode.ExpressionClosureType:
+		closureExpr := val.EncapsulatedValue().(*customdecode.ExpressionClosure)
+		builder.fromExpr(closureExpr.Expression)
+	case typ == customdecode.ExpressionType:
+		builder.fromExpr(customdecode.ExpressionFromVal(val))
+	case val.IsNull():
+		builder.add(ident("null", 0))
+	case typ == cty.Bool:
+		if val.True() {
+			builder.add(ident("true", 0))
+		} else {
+			builder.add(ident("false", 0))
+		}
+	case typ == cty.Number:
+		builder.add(num(val.AsBigFloat()))
+	case typ == cty.String:
+		src := renderQuotedString(val.AsString())
+		builder.add(oquote())
+		if len(src) > 0 {
+			builder.add(&hclwrite.Token{
+				Type:  hclsyntax.TokenQuotedLit,
+				Bytes: src,
+			})
+		}
+		builder.add(cquote())
+	case typ.IsMapType() || typ.IsObjectType():
+		builder.add(obrace())
+		if val.LengthInt() > 0 {
+			builder.add(nl())
+		}
+		i := 0
+		for it := val.ElementIterator(); it.Next(); {
+			eKey, eVal := it.Element()
+			if hclsyntax.ValidIdentifier(eKey.AsString()) {
+				builder.add(ident(eKey.AsString(), 0))
+			} else {
+				builder.fromValue(eKey)
+			}
+			builder.add(assign(0))
+			builder.fromValue(eVal)
+			builder.add(nl())
+			i++
+		}
+		builder.add(cbrace())
+	case typ.IsListType() || typ.IsSetType() || typ.IsTupleType():
+		builder.add(obrack())
+		i := 0
+		for it := val.ElementIterator(); it.Next(); {
+			if i > 0 {
+				builder.add(comma())
+			}
+			_, eVal := it.Element()
+			builder.fromValue(eVal)
+			i++
+		}
+		builder.add(cbrack())
+
+	default:
+		panic(errors.E(errors.ErrInternal, "formatting for value type %s is not supported", val.Type().FriendlyName()))
+	}
+}
+
+func renderQuotedString(s string) []byte {
+	buf := make([]byte, 0, len(s))
+	for i, r := range s {
+		switch r {
+		case '\n':
+			buf = append(buf, '\\', 'n')
+		case '\r':
+			buf = append(buf, '\\', 'r')
+		case '\t':
+			buf = append(buf, '\\', 't')
+		case '"':
+			buf = append(buf, '\\', '"')
+		case '\\':
+			buf = append(buf, '\\', '\\')
+		case '$', '%':
+			buf = appendRune(buf, r)
+			remain := s[i+1:]
+			if len(remain) > 0 && remain[0] == '{' {
+				buf = appendRune(buf, r)
+			}
+		default:
+			if !unicode.IsPrint(r) {
+				var fmted string
+				if r < 65536 {
+					fmted = fmt.Sprintf("\\u%04x", r)
+				} else {
+					fmted = fmt.Sprintf("\\U%08x", r)
+				}
+				buf = append(buf, fmted...)
+			} else {
+				buf = appendRune(buf, r)
+			}
+		}
+	}
+	return buf
+}
+
+func appendRune(b []byte, r rune) []byte {
+	l := utf8.RuneLen(r)
+	for i := 0; i < l; i++ {
+		b = append(b, 0)
+	}
+	ch := b[len(b)-l:]
+	utf8.EncodeRune(ch, r)
+	return b
+}

--- a/hcl/eval/eval.go
+++ b/hcl/eval/eval.go
@@ -11,6 +11,7 @@ import (
 	"github.com/zclconf/go-cty/cty/function"
 
 	hhcl "github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
 )
 
 // ErrEval indicates a failure during the evaluation process
@@ -33,6 +34,14 @@ func NewContext(funcs map[string]function.Function) *Context {
 	return &Context{
 		hclctx: hclctx,
 	}
+}
+
+func (c *Context) newChildContext() *Context {
+	child := NewContext(c.hclctx.Functions)
+	for k, v := range c.hclctx.Variables {
+		child.hclctx.Variables[k] = v
+	}
+	return child
 }
 
 // SetNamespace will set the given values inside the given namespace on the
@@ -86,14 +95,37 @@ func (c *Context) Eval(expr hhcl.Expression) (cty.Value, error) {
 
 // PartialEval evaluates only the terramate variable expressions from the list
 // of tokens, leaving all the rest as-is. It returns a modified list of tokens
-// with  no reference to terramate namespaced variables (globals and terramate)
-// and functions (tm_ prefixed functions).
-func (c *Context) PartialEval(expr hhcl.Expression) (hhcl.Expression, error) {
-	newexpr, err := c.partialEval(expr)
+// with  no reference to terramate namespaced variables (global, let and terramate)
+// and tm_ prefixed functions. It also returns a boolean that tells if the returned
+// expression contains any unknown references (variables or functions).
+// It try to reduce the expression to its simplest form, which means expressions
+// with no unknowns are evaluated down to literals.
+func (c *Context) PartialEval(expr hhcl.Expression) (hhcl.Expression, bool, error) {
+	newexpr, hasUnknowns, err := c.partialEval(expr)
 	if err != nil {
-		return nil, errors.E(ErrPartial, err)
+		return nil, false, errors.E(ErrPartial, err)
 	}
-	return newexpr, nil
+
+	if hasUnknowns {
+		return newexpr, hasUnknowns, nil
+	}
+	switch newexpr.(type) {
+	case *hclsyntax.LiteralValueExpr, *hclsyntax.TemplateExpr, *hclsyntax.TemplateWrapExpr:
+		// NOTE(i4k): Template*Expr are also kept because we support an special
+		// HEREDOC handling detection and then if evaluated it will be converted
+		// to plain quoted strings.
+		return newexpr, hasUnknowns, nil
+	}
+	var evaluated cty.Value
+	evaluated, err = c.Eval(newexpr)
+	if err != nil {
+		return nil, false, errors.E(ErrPartial, err)
+	}
+	newexpr = &hclsyntax.LiteralValueExpr{
+		Val:      evaluated,
+		SrcRange: newexpr.Range(),
+	}
+	return newexpr, hasUnknowns, nil
 }
 
 // Copy the eval context.

--- a/hcl/eval/partial_eval.go
+++ b/hcl/eval/partial_eval.go
@@ -16,18 +16,17 @@ import (
 
 // Errors returned when doing partial evaluation.
 const (
-	ErrPartial             errors.Kind = "partial evaluation failed"
-	ErrInterpolation       errors.Kind = "interpolation failed"
-	ErrForExprDisallowEval errors.Kind = "`for` expression disallow globals/terramate variables"
+	ErrPartial       errors.Kind = "partial evaluation failed"
+	ErrInterpolation errors.Kind = "interpolation failed"
 )
 
-func (c *Context) partialEval(expr hhcl.Expression) (newexpr hhcl.Expression, err error) {
+func (c *Context) partialEval(expr hhcl.Expression) (newexpr hhcl.Expression, hasUnknowns bool, err error) {
 	switch e := expr.(type) {
 	case *ast.CloneExpression:
 		cloned := ast.CloneExpr(e.Expression)
 		return c.partialEval(cloned)
 	case *hclsyntax.LiteralValueExpr:
-		return expr, nil
+		return expr, false, nil
 	case *hclsyntax.UnaryOpExpr:
 		return c.partialEvalUnaryOp(e)
 	case *hclsyntax.BinaryOpExpr:
@@ -58,241 +57,411 @@ func (c *Context) partialEval(expr hhcl.Expression) (newexpr hhcl.Expression, er
 		return c.partialEvalRelTrav(e)
 	case *hclsyntax.ParenthesesExpr:
 		return c.partialEvalParenExpr(e)
+	case *hclsyntax.AnonSymbolExpr:
+		return expr, false, nil
 	default:
 		panic(fmt.Sprintf("not implemented %T", expr))
 	}
 }
 
-func (c *Context) partialEvalTemplate(tmpl *hclsyntax.TemplateExpr) (*hclsyntax.TemplateExpr, error) {
+func (c *Context) partialEvalTemplate(tmpl *hclsyntax.TemplateExpr) (*hclsyntax.TemplateExpr, bool, error) {
+	hasUnknowns := false
 	for i, part := range tmpl.Parts {
-		newexpr, err := c.partialEval(part)
+		newexpr, partHasUnknowns, err := c.partialEval(part)
 		if err != nil {
-			return nil, err
+			return nil, false, err
 		}
+		hasUnknowns = hasUnknowns || partHasUnknowns
 		tmpl.Parts[i] = asSyntax(newexpr)
 	}
-	return tmpl, nil
+	return tmpl, hasUnknowns, nil
 }
 
-func (c *Context) partialEvalTmplWrap(wrap *hclsyntax.TemplateWrapExpr) (hhcl.Expression, error) {
-	newwrap, err := c.partialEval(wrap.Wrapped)
+func (c *Context) partialEvalTmplWrap(wrap *hclsyntax.TemplateWrapExpr) (hhcl.Expression, bool, error) {
+	newwrap, hasUnknowns, err := c.partialEval(wrap.Wrapped)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 	if v, ok := newwrap.(*hclsyntax.LiteralValueExpr); ok {
-		// TODO(fix)
+		// TODO(i4k): check this case.
 		if v.Val.Type() == cty.String && strings.Contains(v.Val.AsString(), "${") {
-			panic(v.Val.AsString())
+			panic(errors.E(errors.ErrInternal, "unexpected case: %s", v.Val.AsString()))
 		}
-		return v, nil
+		return v, hasUnknowns, nil
 	}
 
 	wrap.Wrapped = asSyntax(newwrap)
-	return wrap, nil
+	return wrap, hasUnknowns, nil
 }
 
-func (c *Context) partialEvalTuple(tuple *hclsyntax.TupleConsExpr) (hclsyntax.Expression, error) {
+func (c *Context) partialEvalTuple(tuple *hclsyntax.TupleConsExpr) (hclsyntax.Expression, bool, error) {
+	hasUnknowns := false
 	for i, v := range tuple.Exprs {
-		newexpr, err := c.partialEval(v)
+		newexpr, itHasUnknowns, err := c.partialEval(v)
+		hasUnknowns = hasUnknowns || itHasUnknowns
 		if err != nil {
-			return nil, err
+			return nil, hasUnknowns, err
 		}
 		tuple.Exprs[i] = asSyntax(newexpr)
 	}
-	return tuple, nil
+	return tuple, hasUnknowns, nil
 }
 
-func (c *Context) partialEvalObject(obj *hclsyntax.ObjectConsExpr) (hclsyntax.Expression, error) {
+func (c *Context) partialEvalObject(obj *hclsyntax.ObjectConsExpr) (hclsyntax.Expression, bool, error) {
+	hasUnknowns := false
 	for i, elem := range obj.Items {
-		newkey, err := c.partialEval(elem.KeyExpr)
+		newkey, h1, err := c.partialEval(elem.KeyExpr)
+		hasUnknowns = hasUnknowns || h1
 		if err != nil {
-			return nil, err
+			return nil, false, err
 		}
-		newval, err := c.partialEval(elem.ValueExpr)
+		newval, h2, err := c.partialEval(elem.ValueExpr)
+		hasUnknowns = hasUnknowns || h2
 		if err != nil {
-			return nil, err
+			return nil, false, err
 		}
 		elem.KeyExpr = asSyntax(newkey)
 		elem.ValueExpr = asSyntax(newval)
 		obj.Items[i] = elem
 	}
-	return obj, nil
+	return obj, hasUnknowns, nil
 }
 
-func (c *Context) partialEvalFunc(funcall *hclsyntax.FunctionCallExpr) (hhcl.Expression, error) {
+func (c *Context) partialEvalFunc(funcall *hclsyntax.FunctionCallExpr) (hhcl.Expression, bool, error) {
 	if strings.HasPrefix(funcall.Name, "tm_") {
 		val, err := c.Eval(funcall)
 		if err != nil {
-			return nil, err
+			return nil, false, err
 		}
 		return &hclsyntax.LiteralValueExpr{
 			Val:      val,
 			SrcRange: funcall.Range(),
-		}, nil
+		}, false, nil
 	}
+
+	// hasUnknowns is true because the function is not tm_ prefixed
+
 	for i, arg := range funcall.Args {
-		newexpr, err := c.partialEval(arg)
+		newexpr, _, err := c.partialEval(arg)
 		if err != nil {
-			return nil, err
+			return nil, true, err
 		}
 		funcall.Args[i] = asSyntax(newexpr)
 	}
-	return funcall, nil
+	return funcall, true, nil
 }
 
-func (c *Context) partialEvalIndex(index *hclsyntax.IndexExpr) (hhcl.Expression, error) {
-	newcol, err := c.partialEval(index.Collection)
+func (c *Context) partialEvalIndex(index *hclsyntax.IndexExpr) (hhcl.Expression, bool, error) {
+	hasUnknowns := false
+	newcol, h1, err := c.partialEval(index.Collection)
 	if err != nil {
-		return nil, err
-	}
-	newkey, err := c.partialEval(index.Key)
-	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 	index.Collection = asSyntax(newcol)
-	index.Key = asSyntax(newkey)
-
-	if c.hasUnknownVars(index) {
-		return index, nil
+	hasUnknowns = h1
+	newkey, h2, err := c.partialEval(index.Key)
+	hasUnknowns = hasUnknowns || h2
+	if err != nil {
+		return nil, false, err
 	}
-
+	index.Key = asSyntax(newkey)
+	if hasUnknowns {
+		return index, hasUnknowns, nil
+	}
 	val, err := c.Eval(index)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 	return &hclsyntax.LiteralValueExpr{
 		Val:      val,
 		SrcRange: index.SrcRange,
-	}, nil
+	}, false, nil
 }
 
-func (c *Context) partialEvalSplat(expr *hclsyntax.SplatExpr) (hhcl.Expression, error) {
-	newsrc, err := c.partialEval(expr.Source)
+func (c *Context) partialEvalSplat(expr *hclsyntax.SplatExpr) (hhcl.Expression, bool, error) {
+	newsrc, hasUnknowns, err := c.partialEval(expr.Source)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 	expr.Source = asSyntax(newsrc)
-	if c.hasUnknownVars(expr.Source) {
-		return expr, nil
+	if hasUnknowns {
+		return expr, hasUnknowns, nil
 	}
-	if c.hasUnknownVars(expr.Each) {
-		return expr, nil
+	newEach, hasUnknowns, err := c.partialEval(expr.Each)
+	if err != nil {
+		return nil, false, err
+	}
+	expr.Each = asSyntax(newEach)
+	if hasUnknowns {
+		return expr, hasUnknowns, nil
 	}
 	val, err := c.Eval(expr)
 	if err != nil {
-		// this can happen in the case of using funcalls not prefixed with tm_
-		return expr, nil
+		return expr, false, err
 	}
 	return &hclsyntax.LiteralValueExpr{
 		Val:      val,
 		SrcRange: expr.Range(),
-	}, nil
+	}, false, nil
 }
 
-func (c *Context) hasUnknownVars(expr hclsyntax.Expression) bool {
-	for _, namespace := range expr.Variables() {
-		if !c.HasNamespace(namespace.RootName()) {
-			return true
-		}
-	}
-	return false
-}
-
-func (c *Context) hasTerramateVars(expr hclsyntax.Expression) bool {
-	for _, namespace := range expr.Variables() {
-		if c.HasNamespace(namespace.RootName()) {
-			return true
-		}
-	}
-	return false
-}
-
-func (c *Context) partialEvalObjectKey(key *hclsyntax.ObjectConsKeyExpr) (hhcl.Expression, error) {
+func (c *Context) partialEvalObjectKey(key *hclsyntax.ObjectConsKeyExpr) (hhcl.Expression, bool, error) {
 	var (
-		wrapped hhcl.Expression
-		err     error
+		err         error
+		wrapped     hhcl.Expression
+		hasUnknowns bool
 	)
 
 	switch vexpr := key.Wrapped.(type) {
 	case *hclsyntax.ScopeTraversalExpr:
-		wrapped, err = c.partialEvalScopeTrav(vexpr, partialEvalOption{
+		wrapped, hasUnknowns, err = c.partialEvalScopeTrav(vexpr, partialEvalOption{
 			// back-compatibility with old partial evaluator.
 			// global = 1
 			forbidRootEval: true,
 		})
 	case *hclsyntax.ParenthesesExpr:
-		wrapped, err = c.partialEvalParenExpr(vexpr, partialEvalOption{
+		wrapped, hasUnknowns, err = c.partialEvalParenExpr(vexpr, partialEvalOption{
 			// back-compatibility with old partial evaluator.
 			// (global) = 1
-			forbidRootEval: true,
+			forbidRootEval: false,
 		})
 	default:
-		wrapped, err = c.partialEval(key.Wrapped)
+		wrapped, hasUnknowns, err = c.partialEval(key.Wrapped)
 	}
 
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 
 	key.Wrapped = asSyntax(wrapped)
-	return key, nil
+	return key, hasUnknowns, nil
 }
 
-func (c *Context) partialEvalForExpr(forExpr *hclsyntax.ForExpr) (hhcl.Expression, error) {
-	for _, expr := range []hclsyntax.Expression{
-		forExpr.KeyExpr,
-		forExpr.ValExpr,
-		forExpr.CollExpr,
-		forExpr.CondExpr,
-	} {
-		if expr != nil && c.hasTerramateVars(forExpr.CollExpr) {
-			return nil, errors.E(ErrForExprDisallowEval, "evaluating expression: %s", ast.TokensForExpression(forExpr.CollExpr).Bytes())
+func (c *Context) partialEvalForExpr(forExpr *hclsyntax.ForExpr) (hhcl.Expression, bool, error) {
+	resExpr, hasUnknowns, err := c.partialEval(forExpr.CollExpr)
+	if err != nil {
+		return nil, false, err
+	}
+	forExpr.CollExpr = asSyntax(resExpr)
+	if !hasUnknowns {
+		resExpr, hasUnknowns, err = c.evalForLoop(forExpr)
+		if err != nil {
+			return nil, false, err
+		}
+		if !hasUnknowns {
+			return resExpr, false, nil
 		}
 	}
 
-	return forExpr, nil
+	if forExpr.KeyExpr != nil {
+		resExpr, _, err = c.partialEval(forExpr.KeyExpr)
+		if err != nil {
+			return nil, false, err
+		}
+		forExpr.KeyExpr = asSyntax(resExpr)
+	}
+
+	if forExpr.ValExpr != nil {
+		resExpr, _, err := c.partialEval(forExpr.ValExpr)
+		if err != nil {
+			return nil, false, err
+		}
+		forExpr.ValExpr = asSyntax(resExpr)
+	}
+
+	if forExpr.CondExpr != nil {
+		resExpr, _, err := c.partialEval(forExpr.CondExpr)
+		if err != nil {
+			return nil, false, err
+		}
+		forExpr.CondExpr = asSyntax(resExpr)
+	}
+
+	return forExpr, true, nil
 }
 
-func (c *Context) partialEvalBinOp(binop *hclsyntax.BinaryOpExpr) (hhcl.Expression, error) {
-	lhs, err := c.partialEval(binop.LHS)
-	if err != nil {
-		return nil, err
+func (c *Context) evalForLoop(forExpr *hclsyntax.ForExpr) (hhcl.Expression, bool, error) {
+	if forExpr.KeyExpr != nil {
+		return c.evalForObjectLoop(forExpr)
 	}
-	rhs, err := c.partialEval(binop.RHS)
+	return c.evalForListLoop(forExpr)
+}
+
+func (c *Context) evalForObjectLoop(forExpr *hclsyntax.ForExpr) (hhcl.Expression, bool, error) {
+	res := make(map[string]cty.Value)
+	coll, err := c.Eval(forExpr.CollExpr)
 	if err != nil {
-		return nil, err
+		return nil, false, err
+	}
+	if !coll.CanIterateElements() {
+		return nil, false, errors.E("for-expr with non-iterable collection", forExpr.CollExpr.Range())
+	}
+	iterator := coll.ElementIterator()
+	for iterator.Next() {
+		k, v := iterator.Element()
+		childCtx := c.newChildContext()
+		childCtx.hclctx.Variables[forExpr.KeyVar] = k
+		childCtx.hclctx.Variables[forExpr.ValVar] = v
+
+		if forExpr.CondExpr != nil {
+			condExpr, hasUnknowns, err := childCtx.partialEval(&ast.CloneExpression{
+				Expression: forExpr.CondExpr,
+			})
+			if err != nil {
+				return nil, false, err
+			}
+			if hasUnknowns {
+				return forExpr, hasUnknowns, nil
+			}
+			condVal, err := childCtx.Eval(condExpr)
+			if err != nil {
+				return nil, false, err
+			}
+			if condVal.Type() != cty.Bool {
+				return nil, false, errors.E("condition is not a boolean but %s", condVal.Type().FriendlyName())
+			}
+			if condVal.False() {
+				continue
+			}
+		}
+		resKeyExpr, hasUnknowns, err := childCtx.partialEval(&ast.CloneExpression{
+			Expression: forExpr.KeyExpr,
+		})
+		if err != nil {
+			return nil, false, err
+		}
+
+		if hasUnknowns {
+			return forExpr, hasUnknowns, nil
+		}
+
+		resKeyVal, err := childCtx.Eval(resKeyExpr)
+		if err != nil {
+			return nil, false, err
+		}
+		if resKeyVal.Type() != cty.String {
+			return nil, false, errors.E("object key must be a string", forExpr.KeyExpr.Range())
+		}
+		resValExpr, hasUnknowns, err := childCtx.partialEval(&ast.CloneExpression{
+			Expression: forExpr.ValExpr,
+		})
+		if err != nil {
+			return nil, false, err
+		}
+		if hasUnknowns {
+			return forExpr, hasUnknowns, nil
+		}
+		resVal, err := childCtx.Eval(resValExpr)
+		if err != nil {
+			return nil, false, err
+		}
+		res[resKeyVal.AsString()] = resVal
+	}
+	return &hclsyntax.LiteralValueExpr{
+		Val:      cty.ObjectVal(res),
+		SrcRange: forExpr.SrcRange,
+	}, false, nil
+}
+
+func (c *Context) evalForListLoop(forExpr *hclsyntax.ForExpr) (hhcl.Expression, bool, error) {
+	var res []cty.Value
+	coll, err := c.Eval(forExpr.CollExpr)
+	if err != nil {
+		return nil, false, err
+	}
+	if !coll.CanIterateElements() {
+		return nil, false, errors.E("for-expr with non-iterable collection", forExpr.CollExpr.Range())
+	}
+	iterator := coll.ElementIterator()
+	for iterator.Next() {
+		k, v := iterator.Element()
+		childCtx := c.newChildContext()
+		childCtx.hclctx.Variables[forExpr.KeyVar] = k
+		childCtx.hclctx.Variables[forExpr.ValVar] = v
+
+		if forExpr.CondExpr != nil {
+			condExpr, hasUnknowns, err := childCtx.partialEval(&ast.CloneExpression{
+				Expression: forExpr.CondExpr,
+			})
+			if err != nil {
+				return nil, false, err
+			}
+			if hasUnknowns {
+				return forExpr, hasUnknowns, nil
+			}
+			condVal, err := childCtx.Eval(condExpr)
+			if err != nil {
+				return nil, false, err
+			}
+			if condVal.Type() != cty.Bool {
+				return nil, false, errors.E("condition is not a boolean but %s", condVal.Type().FriendlyName())
+			}
+			if condVal.False() {
+				continue
+			}
+		}
+
+		resValExpr, hasUnknowns, err := childCtx.partialEval(&ast.CloneExpression{
+			Expression: forExpr.ValExpr,
+		})
+		if err != nil {
+			return nil, false, err
+		}
+		if hasUnknowns {
+			return forExpr, hasUnknowns, nil
+		}
+		resVal, err := childCtx.Eval(resValExpr)
+		if err != nil {
+			return nil, false, err
+		}
+		res = append(res, resVal)
+	}
+	return &hclsyntax.LiteralValueExpr{
+		Val:      cty.TupleVal(res),
+		SrcRange: forExpr.SrcRange,
+	}, false, nil
+}
+
+func (c *Context) partialEvalBinOp(binop *hclsyntax.BinaryOpExpr) (hhcl.Expression, bool, error) {
+	lhs, h1, err := c.partialEval(binop.LHS)
+	if err != nil {
+		return nil, false, err
+	}
+	rhs, h2, err := c.partialEval(binop.RHS)
+	if err != nil {
+		return nil, false, err
 	}
 	binop.LHS = asSyntax(lhs)
 	binop.RHS = asSyntax(rhs)
-	return binop, nil
+	return binop, h1 || h2, nil
 }
 
-func (c *Context) partialEvalUnaryOp(unary *hclsyntax.UnaryOpExpr) (hhcl.Expression, error) {
-	val, err := c.partialEval(unary.Val)
+func (c *Context) partialEvalUnaryOp(unary *hclsyntax.UnaryOpExpr) (hhcl.Expression, bool, error) {
+	val, hasUnknowns, err := c.partialEval(unary.Val)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 	unary.Val = asSyntax(val)
-	return unary, nil
+	return unary, hasUnknowns, nil
 }
 
-func (c *Context) partialEvalCondExpr(cond *hclsyntax.ConditionalExpr) (hhcl.Expression, error) {
-	newcond, err := c.partialEval(cond.Condition)
+func (c *Context) partialEvalCondExpr(cond *hclsyntax.ConditionalExpr) (hhcl.Expression, bool, error) {
+	newcond, h1, err := c.partialEval(cond.Condition)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
-	newtrue, err := c.partialEval(cond.TrueResult)
+	newtrue, h2, err := c.partialEval(cond.TrueResult)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
-	newfalse, err := c.partialEval(cond.FalseResult)
+	newfalse, h3, err := c.partialEval(cond.FalseResult)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 	cond.Condition = asSyntax(newcond)
 	cond.TrueResult = asSyntax(newtrue)
 	cond.FalseResult = asSyntax(newfalse)
-	return cond, nil
+	return cond, h1 || h2 || h3, nil
 }
 
 type partialEvalOption struct {
@@ -300,35 +469,33 @@ type partialEvalOption struct {
 	forbidRootEval bool
 }
 
-func (c *Context) partialEvalScopeTrav(scope *hclsyntax.ScopeTraversalExpr, opts ...partialEvalOption) (hclsyntax.Expression, error) {
+func (c *Context) partialEvalScopeTrav(scope *hclsyntax.ScopeTraversalExpr, opts ...partialEvalOption) (hclsyntax.Expression, bool, error) {
 	assertPartialExprOpt(opts)
 	ns, ok := scope.Traversal[0].(hhcl.TraverseRoot)
 	if !ok {
-		return scope, nil
-	}
-	if !c.HasNamespace(ns.Name) {
-		return scope, nil
+		return scope, false, nil
 	}
 	forbidRootEval := false
 	if len(opts) == 1 {
 		forbidRootEval = opts[0].forbidRootEval
 	}
 	if len(scope.Traversal) == 1 && forbidRootEval {
-		return scope, nil
+		return scope, false, nil
 	}
-
+	if !c.HasNamespace(ns.Name) {
+		return scope, true, nil
+	}
 	val, err := c.Eval(scope)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
-
 	return &hclsyntax.LiteralValueExpr{
 		Val:      val,
 		SrcRange: scope.SrcRange,
-	}, nil
+	}, false, nil
 }
 
-func (c *Context) partialEvalParenExpr(paren *hclsyntax.ParenthesesExpr, opts ...partialEvalOption) (hhcl.Expression, error) {
+func (c *Context) partialEvalParenExpr(paren *hclsyntax.ParenthesesExpr, opts ...partialEvalOption) (hhcl.Expression, bool, error) {
 	assertPartialExprOpt(opts)
 	forbidRootEval := false
 	if len(opts) == 1 {
@@ -336,42 +503,43 @@ func (c *Context) partialEvalParenExpr(paren *hclsyntax.ParenthesesExpr, opts ..
 	}
 
 	var (
-		newexpr hhcl.Expression
-		err     error
+		hasUnknowns bool
+		newexpr     hhcl.Expression
+		err         error
 	)
 
 	if scope, ok := paren.Expression.(*hclsyntax.ScopeTraversalExpr); ok && forbidRootEval {
-		newexpr, err = c.partialEvalScopeTrav(scope, opts...)
+		newexpr, hasUnknowns, err = c.partialEvalScopeTrav(scope, opts...)
 	} else {
-		newexpr, err = c.partialEval(paren.Expression)
+		newexpr, hasUnknowns, err = c.partialEval(paren.Expression)
 	}
 
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 
 	paren.Expression = asSyntax(newexpr)
-	return paren, nil
+	return paren, hasUnknowns, nil
 }
 
-func (c *Context) partialEvalRelTrav(rel *hclsyntax.RelativeTraversalExpr) (hhcl.Expression, error) {
-	newsrc, err := c.partialEval(rel.Source)
+func (c *Context) partialEvalRelTrav(rel *hclsyntax.RelativeTraversalExpr) (hhcl.Expression, bool, error) {
+	newsrc, hasUnknowns, err := c.partialEval(rel.Source)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 	rel.Source = asSyntax(newsrc)
-	if c.hasUnknownVars(rel) {
-		return rel, nil
+	if hasUnknowns {
+		return rel, hasUnknowns, nil
 	}
 
 	val, err := c.Eval(rel)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 	return &hclsyntax.LiteralValueExpr{
 		Val:      val,
 		SrcRange: rel.SrcRange,
-	}, nil
+	}, false, nil
 }
 
 func asSyntax(expr hhcl.Expression) hclsyntax.Expression {

--- a/hcl/eval/partial_eval_bench_test.go
+++ b/hcl/eval/partial_eval_bench_test.go
@@ -99,7 +99,7 @@ func BenchmarkPartialEvalComplex(b *testing.B) {
 		if diags.HasErrors() {
 			b.Fatalf(diags.Error())
 		}
-		_, err := ctx.PartialEval(expr)
+		_, _, err := ctx.PartialEval(expr)
 		if err != nil {
 			b.Fatal(err.Error())
 		}
@@ -118,7 +118,7 @@ func BenchmarkPartialEvalSmallString(b *testing.B) {
 		if diags.HasErrors() {
 			b.Fatalf(diags.Error())
 		}
-		_, err := ctx.PartialEval(expr)
+		_, _, err := ctx.PartialEval(expr)
 		if err != nil {
 			b.Fatal(err.Error())
 		}
@@ -137,7 +137,7 @@ func BenchmarkPartialEvalHugeString(b *testing.B) {
 		if diags.HasErrors() {
 			b.Fatalf(diags.Error())
 		}
-		_, err := ctx.PartialEval(expr)
+		_, _, err := ctx.PartialEval(expr)
 		if err != nil {
 			b.Fatal(err.Error())
 		}
@@ -156,7 +156,7 @@ func BenchmarkPartialEvalHugeInterpolatedString(b *testing.B) {
 		if diags.HasErrors() {
 			b.Fatalf(diags.Error())
 		}
-		_, err := ctx.PartialEval(expr)
+		_, _, err := ctx.PartialEval(expr)
 		if err != nil {
 			b.Fatal(err.Error())
 		}
@@ -180,7 +180,7 @@ func BenchmarkPartialEvalObject(b *testing.B) {
 		if diags.HasErrors() {
 			b.Fatalf(diags.Error())
 		}
-		_, err := ctx.PartialEval(expr)
+		_, _, err := ctx.PartialEval(expr)
 		if err != nil {
 			b.Fatal(err.Error())
 		}

--- a/hcl/eval/partial_eval_test.go
+++ b/hcl/eval/partial_eval_test.go
@@ -7,10 +7,10 @@ import (
 	"os"
 	"testing"
 
+	"github.com/google/go-cmp/cmp"
 	"github.com/hashicorp/hcl/v2"
 	"github.com/hashicorp/hcl/v2/hclsyntax"
 	"github.com/hashicorp/hcl/v2/hclwrite"
-	"github.com/madlambda/spells/assert"
 	"github.com/terramate-io/terramate/errors"
 	"github.com/terramate-io/terramate/hcl/ast"
 	"github.com/terramate-io/terramate/hcl/eval"
@@ -23,9 +23,10 @@ func TestPartialEval(t *testing.T) {
 	t.Parallel()
 
 	type testcase struct {
-		expr    string
-		want    string
-		wantErr error
+		expr        string
+		want        string
+		hasUnknowns bool
+		wantErr     error
 	}
 
 	for _, tc := range []testcase{
@@ -34,6 +35,11 @@ func TestPartialEval(t *testing.T) {
 		},
 		{
 			expr: `!1`,
+			//  Unsuitable value for unary operand: bool required.
+			wantErr: errors.E(eval.ErrPartial),
+		},
+		{
+			expr: `true`,
 		},
 		{
 			expr: `1.5`,
@@ -52,15 +58,20 @@ EOT
 `,
 		},
 		{
-			expr: `"test ${unknown.val}"`,
+			expr:        `"test ${unknown.val}"`,
+			hasUnknowns: true,
 		},
 		{
 			expr: `global.string`,
 			want: `"terramate"`,
 		},
 		{
-			expr: `!global.number`,
-			want: `!10`,
+			expr:    `!global.number`,
+			wantErr: errors.E(eval.ErrPartial),
+		},
+		{
+			expr: `!global.falsy`,
+			want: `true`,
 		},
 		{
 			expr: `"test ${global.number}"`,
@@ -85,7 +96,12 @@ EOT
 		},
 		{
 			expr: `[!global.number]`,
-			want: `[!10]`,
+			// Unsuitable value for unary operand: bool required.
+			wantErr: errors.E(eval.ErrPartial),
+		},
+		{
+			expr: `[!global.falsy]`,
+			want: `[true]`,
 		},
 		{
 			expr: `{
@@ -128,31 +144,39 @@ EOT
 		},
 		{
 			expr: `1+1`,
+			want: `2`,
 		},
 		{
-			expr: `1+data.val`,
+			expr:        `1+data.val`,
+			hasUnknowns: true,
 		},
 		{
-			expr: `data.val1+1000+data.val2`,
+			expr:        `data.val1+1000+data.val2`,
+			hasUnknowns: true,
 		},
 		{
 			expr: `1+global.number`,
-			want: `1+10`,
+			want: `11`,
 		},
 		{
 			expr: `global.string+global.number`,
-			want: `"terramate"+10`,
+			// Unsuitable value for left operand: a number is required.
+			wantErr: errors.E(eval.ErrPartial),
+		},
+		{
+			expr: `global.number+global.number`,
+			want: `20`,
 		},
 		{
 			expr: `[1+global.number]`,
-			want: `[1+10]`,
+			want: `[11]`,
 		},
 		{
 			expr: `{
 				a = 1+global.number
 			}`,
 			want: `{
-				a = 1+10
+				a = 11
 			}`,
 		},
 		{
@@ -160,7 +184,7 @@ EOT
 				global.string = 1
 			}`,
 			want: `{
-				"terramate" = 1	
+				terramate = 1	
 			}`,
 		},
 		{
@@ -168,7 +192,7 @@ EOT
 				global.obj.b[0] = 1
 			}`,
 			want: `{
-				"terramate" = 1	
+				terramate = 1	
 			}`,
 		},
 		{
@@ -183,9 +207,17 @@ EOT
 			expr: `{
 				(global) = 1
 			}`,
-			want: `{
-				(global) = 1	
+			// Can't use this value as a key: string required
+			wantErr: errors.E(eval.ErrPartial),
+		},
+		{
+			expr: `{
+				(aws.vpc.id) = 1
 			}`,
+			want: `{
+				(aws.vpc.id) = 1	
+			}`,
+			hasUnknowns: true,
 		},
 		{
 			expr: `{
@@ -194,10 +226,12 @@ EOT
 			want: `{
 				(iter) = 1	
 			}`,
+			hasUnknowns: true,
 		},
 		{
 			expr: `global`,
 			want: `{
+				falsy  = false
 				list   = [0, 1, 2, 3]
 				number = 10
 				obj = {
@@ -206,14 +240,15 @@ EOT
 				}
 				string  = "terramate"
 				strings = ["terramate", "is", "fun"]
-			}`,
+				truer   = true
+			  }`,
 		},
 		{
 			expr: `{
 				(global.string) = 1
 			}`,
 			want: `{
-				("terramate") = 1
+				terramate = 1
 			}`,
 		},
 		{
@@ -221,7 +256,7 @@ EOT
 				tm_upper(global.string) = 1
 			}`,
 			want: `{
-				"TERRAMATE" = 1
+				TERRAMATE = 1
 			}`,
 		},
 		{
@@ -229,7 +264,7 @@ EOT
 				(tm_upper(global.string)) = 1
 			}`,
 			want: `{
-				("TERRAMATE") = 1
+				TERRAMATE = 1
 			}`,
 		},
 		{
@@ -239,6 +274,7 @@ EOT
 			want: `{
 				upper("terramate") = 1
 			}`,
+			hasUnknowns: true,
 		},
 		{
 			expr: `{
@@ -247,6 +283,7 @@ EOT
 			want: `{
 				upper("a") = 1
 			}`,
+			hasUnknowns: true,
 		},
 		{
 			expr: `{
@@ -255,6 +292,7 @@ EOT
 			want: `{
 				(upper("a")) = 1
 			}`,
+			hasUnknowns: true,
 		},
 		{
 			expr: `{
@@ -263,30 +301,123 @@ EOT
 			want: `{
 				a.b.c = 1
 			}`,
+			hasUnknowns: true,
 		},
 		{
-			expr: `funcall()`,
+			expr:        `funcall()`,
+			hasUnknowns: true,
 		},
 		{
-			expr: `funcall(1, 2, data.val, "test", {}, [1, 2])`,
+			expr:        `funcall(1, 2, data.val, "test", {}, [1, 2])`,
+			hasUnknowns: true,
 		},
 		{
-			expr: `funcall(1, 2, global.number, 3)`,
-			want: `funcall(1, 2, 10, 3)`,
+			expr:        `funcall(1, 2, global.number, 3)`,
+			want:        `funcall(1, 2, 10, 3)`,
+			hasUnknowns: true,
 		},
 		{
-			expr: `[for v in val : v if v+1 == data.something]`,
+			expr:        `[for v in val : v if v+1 == data.something]`,
+			hasUnknowns: true,
 		},
 		{
-			expr: `{for k, v in val : k => funcall(v) if v+1 == data.something}`,
+			expr:        `{for k, v in val : k => funcall(v) if v+1 == data.something}`,
+			hasUnknowns: true,
 		},
 		{
-			expr:    `[for v in global.list : v if v+data == otherdata.val]`,
-			wantErr: errors.E(eval.ErrForExprDisallowEval),
+			expr:        `[for v in global.list : v if v+data == otherdata.val]`,
+			want:        `[for v in [0, 1, 2, 3] : v if v + data == otherdata.val]`,
+			hasUnknowns: true,
 		},
 		{
-			expr:    `[for v in global.strings : upper(v)]`,
-			wantErr: errors.E(eval.ErrForExprDisallowEval),
+			expr: `[for v in global.strings : tm_upper(v)]`,
+			want: `["TERRAMATE", "IS", "FUN"]`,
+		},
+		{
+			expr: `{for k, v in global.obj : tm_upper(k) => v if k == "a"}`,
+			want: `{
+				A = 0
+			}`,
+		},
+		{
+			expr: `{for k, v in global.obj : k => v+100 if k == "a"}`,
+			want: `{
+				a = 100
+			}`,
+		},
+		{
+			// loop cannot be evaluated because `coll` refers to unknowns
+			expr: `{for k, v in unknown_func(global.obj) : k => v if var.data}`,
+			want: `{for k, v in unknown_func({
+				a = 0
+				b = ["terramate"]
+			  }) : k => v if var.data}`,
+			hasUnknowns: true,
+		},
+		{
+			// loop cannot be evaluated because `if ...` refers to unknowns
+			expr: `{for k, v in global.obj : k => v if var.data}`,
+			want: `{for k, v in {
+				a = 0
+				b = ["terramate"]
+			  } : k => v if var.data}`,
+			hasUnknowns: true,
+		},
+		{
+			// loop cannot be evaluated because `if ...` refers to unknowns function
+			expr: `{for k, v in global.obj : k => v if unknown_func(v)}`,
+			want: `{for k, v in {
+				a = 0
+				b = ["terramate"]
+			  } : k => v if unknown_func(v)}`,
+			hasUnknowns: true,
+		},
+		{
+			// loop cannot be reduced because `key` refers to unknowns function
+			expr: `{for k, v in global.obj : unknown_func(k) => v}`,
+			want: `{for k, v in {
+				a = 0
+				b = ["terramate"]
+			  } : unknown_func(k) => v}`,
+			hasUnknowns: true,
+		},
+		{
+			// loop cannot be reduced because `value` refers to unknowns function
+			expr: `{for k, v in global.obj : k => unknown_func(v)}`,
+			want: `{for k, v in {
+				a = 0
+				b = ["terramate"]
+			  } : k => unknown_func(v)}`,
+			hasUnknowns: true,
+		},
+		{
+			expr: `{for k, v in global.obj : unknown => v if k == "a"}`,
+			want: `{ for k, v in {
+				a = 0
+				b = ["terramate"]
+			  } : unknown => v if k == "a" }`,
+			hasUnknowns: true,
+		},
+		{
+			expr:    `{for k, v in global.obj : v => v+1 if k == "a"}`,
+			wantErr: errors.E(eval.ErrPartial), // v is a number
+		},
+		{
+			expr:    `{for k, v in global.obj : v => v+1 if k == "a"}`,
+			wantErr: errors.E(eval.ErrPartial), // v is a number
+		},
+		{
+			expr:        `[for v in tm_concat(global.list, [4, 5, 6]) : v if v+data == otherdata.val]`,
+			want:        `[for v in [0, 1, 2, 3, 4, 5, 6] : v if v + data == otherdata.val]`,
+			hasUnknowns: true,
+		},
+		{
+			expr: `{for k in data.something : k => global.obj[k]}`,
+			want: `{for k in data.something : k => {
+				a = 0
+				b = ["terramate"]
+			  }[k]}`,
+			hasUnknowns: true,
 		},
 		{
 			expr: `global.list[0]`,
@@ -302,14 +433,35 @@ EOT
 		},
 		{
 			expr: `true?[]:{}`,
+			// different branch types
+			wantErr: errors.E(eval.ErrPartial),
+		},
+		{
+			expr: `true?"truth":"fake"`,
+			want: `"truth"`,
+		},
+		{
+			expr: `false?"truth":"fake"`,
+			want: `"fake"`,
 		},
 		{
 			expr: `global.number == 10?global.string:global.list`,
-			want: `10 == 10 ? "terramate" : [0, 1, 2, 3]`,
+			// err: eval expression: The true and false result expressions must have consistent types. The 'true' value is string, but the 'false' value is list of number.
+			wantErr: errors.E(eval.ErrPartial),
+		},
+		{
+			expr: `global.number == 10?global.string:"test"`,
+			want: `"terramate"`,
 		},
 		{
 			expr: `[0, 1, 2][0]`,
 			want: `0`,
+		},
+		// fuzzer case
+		{
+			expr:        `A.0[0.*]`,
+			want:        `A[0][[0]]`,
+			hasUnknowns: true,
 		},
 		{
 			expr: `[0, 1, 2][1]`,
@@ -332,7 +484,7 @@ EOT
 	} {
 		tc := tc
 		t.Run(tc.expr, func(t *testing.T) {
-			t.Parallel()
+			//t.Parallel()
 			ctx := eval.NewContext(stdlib.Functions(os.TempDir()))
 			ctx.SetNamespace("global", map[string]cty.Value{
 				"number": cty.NumberIntVal(10),
@@ -352,12 +504,14 @@ EOT
 					"a": cty.NumberIntVal(0),
 					"b": cty.ListVal([]cty.Value{cty.StringVal("terramate")}),
 				}),
+				"truer": cty.True,
+				"falsy": cty.False,
 			})
 			expr, diags := hclsyntax.ParseExpression([]byte(tc.expr), "test.hcl", hcl.InitialPos)
 			if diags.HasErrors() {
 				t.Fatalf(diags.Error())
 			}
-			gotExpr, err := ctx.PartialEval(expr)
+			gotExpr, hasUnknowns, err := ctx.PartialEval(expr)
 			errtest.Assert(t, err, tc.wantErr)
 			if tc.wantErr != nil {
 				return
@@ -367,7 +521,16 @@ EOT
 				want = tc.want
 			}
 			got := ast.TokensForExpression(gotExpr)
-			assert.EqualStrings(t, string(hclwrite.Format([]byte(want))), string(hclwrite.Format(got.Bytes())))
+			wantFormatted := string(hclwrite.Format([]byte(want)))
+			gotFormatted := string(hclwrite.Format(got.Bytes()))
+			t.Logf("got:  '%s'", gotFormatted)
+			t.Logf("want: '%s'", wantFormatted)
+			if diff := cmp.Diff(wantFormatted, gotFormatted); diff != "" {
+				t.Fatal(diff)
+			}
+			if hasUnknowns != tc.hasUnknowns {
+				t.Fatalf("hasUnknowns mismatch: got[%t] != want[%t]", hasUnknowns, tc.hasUnknowns)
+			}
 		})
 	}
 }

--- a/hcl/eval/partial_fuzz_test.go
+++ b/hcl/eval/partial_fuzz_test.go
@@ -95,7 +95,7 @@ EOT`,
 		ctx.SetNamespace("global", globals)
 		ctx.SetNamespace("terramate", terramate)
 
-		gotExpr, err := ctx.PartialEval(parsedExpr)
+		gotExpr, _, err := ctx.PartialEval(parsedExpr)
 		if err != nil {
 			return
 		}

--- a/hcl/hcl.go
+++ b/hcl/hcl.go
@@ -322,7 +322,8 @@ type Evaluator interface {
 	// tokens that form the result of the partial evaluation. Any unknown
 	// namespace access are ignored and left as is, while known namespaces
 	// are substituted by its value.
-	PartialEval(hcl.Expression) (hcl.Expression, error)
+	// If any unknowns are found, the method returns hasUnknowns as true.
+	PartialEval(hcl.Expression) (expr hcl.Expression, hasUnknowns bool, err error)
 
 	// SetNamespace adds a new namespace, replacing any with the same name.
 	SetNamespace(name string, values map[string]cty.Value)

--- a/stdlib/ternary.go
+++ b/stdlib/ternary.go
@@ -51,13 +51,14 @@ func evalTernaryBranch(arg cty.Value) (cty.Value, error) {
 	closure := customdecode.ExpressionClosureFromVal(arg)
 
 	ctx := eval.NewContextFrom(closure.EvalContext)
-	newexpr, err := ctx.PartialEval(&ast.CloneExpression{
+	newexpr, _, err := ctx.PartialEval(&ast.CloneExpression{
 		Expression: closure.Expression.(hclsyntax.Expression),
 	})
 	if err != nil {
 		return cty.NilVal, errors.E(err, "evaluating tm_ternary branch")
 	}
 
+	// TODO(i4k): use our own hasUnknowns here.
 	if dependsOnUnknowns(newexpr, closure.EvalContext) {
 		return customdecode.ExpressionVal(newexpr), nil
 	}

